src/Native/Bootstrap/main.cpp

```diff
@@ -318,29 +376,37 @@ int main(int argc, char* argv[])
     }
 #endif // !CPPCODEGEN
 
+    RUNTIME_INITIALIZED = true;
+
 #ifndef CPPCODEGEN
     InitializeModules(osModule, __modules_a, (int)((__modules_z - __modules_a)), (void **)&c_classlibFunctions, _countof(c_classlibFunctions));
 #else // !CPPCODEGEN
     InitializeModules(nullptr, (void**)RtRHeaderWrapper(), 2, nullptr, 0);
 #endif // !CPPCODEGEN
 
-    int retval;
-#ifdef CPPCODEGEN
-    try
+#ifdef CORERT_DLL
+    // Run startup method immediately for a native library
+    return __managed__Startup(0, NULL);
+#else
+    return 0;
+#endif // CORERT_DLL
+}
+
+#ifdef CORERT_DLL
+#if defined(_WIN32)
+extern "C" BOOL WINAPI DllMain(
```

Comment:
Whoever is linking the static library may want to have custom DllMain too. This will make it impossible.

Have you tried using static constructor for the initialization?

```
static struct InitializeRuntimePointerHelper
{
	InitializeRuntimePointerHelper()
	{
    ...
	}
} initializeRuntimePointerHelper;
```

Reply:
@jkotas won't this mean that there'll still be a static dependency on the bootstrapper (e.g. with .NET Native on Windows UWP)? Would this static construct be run when this library is loaded?

Reply:
The dependencies from runtime to the bootstrapper are the problematic ones. If you add the static constructor to the boostrapper, I do not see any problems with dependencies.